### ISSUE TITLE
Support for Firebase Dynamic Links (aka Deep links)

### ISSUE THREAD

**Issue by sethladd** [flutter/flutter#11540]
From a current Flutter user, they wish to have support for Deep Links aka Dynamic Links. See https://firebase.google.com/products/dynamic-links/

**Comment by sethladd** [comment]
"We need a URL that will 1) take the user to a specific part in our app (a chat, for example) if they have the app installed or 2) take a user to the device specific app store and then do 1) after they install the app.
I think this is called Google Play Referrer on Android "

Do we have any examples of how to do this, with Flutter?

**Comment by sethladd** [comment]
See also https://github.com/flutter/flutter/issues/10975 which has some discussion

**Comment by sethladd** [comment]
https://github.com/flutter/flutter/issues/11607 is the request to document this

**Comment by collinjackson** [comment]
There are a few more steps required to implement the Dynamic Links "receive" integration on [iOS](https://firebase.google.com/docs/dynamic-links/ios/receive) and [Android](https://firebase.google.com/docs/dynamic-links/android/receive), like adding a dependency on a compatible Dynamic Links mobile SDK. If we provided a plugin, it could do some steps of this integration automatically, as well as receiving Firebase Dynamic Links and using them to set the initial route. The developer still has to set up an intent filter in the Android manifest and set deepLinkURLScheme on iOS; I don't see an obvious way around that.

Alternatively, devs could do the integration by hand on iOS and Android and set the initial route, and not use a plugin.

The plugin could also provide an API for constructing dynamic links, but I think that can be split off into a separate issue, and that isn't what customer: carnivore is trying to do.

@mit-mit please let me know what you'd like to see happen here and if there's anything I can do to help.

**Comment by Hixie** [comment]
We will probably not get to this in the near future, but it is on our radar for a few months from now.

**Comment by sethladd** [comment]
Our friends at Carnivore implemented an in-house version of this. For now, they are unblocked.

**Comment by bjornbjorn** [comment]
Hi @sethladd - did Carnivore implement an in-house version using Firebase Dynamic Links or just general deep linking?

Either way, any more info / examples on this issue would be highly appreciated! :-)

**Comment by sethladd** [comment]
I've pinged Carnivore, not sure the status. And yes, would love to see this available to everyone!

**Comment by kykl** [comment]
@collinjackson This is exactly what we did as you said it 
'Alternatively, devs could do the integration by hand on iOS and Android and set the initial route, and not use a plugin.''

**Comment by collinjackson** [comment]
I cleared the milestone, I think we should re-triage this issue in light of Carnivore's experience

**Comment by zoechi** [comment]
I also need this for a customer project. I assume I can build something myself that works for now, but to have a proper integration available before the release would be great.

**Comment by Arno-Z** [comment]
Just a friendly bump...

**Comment by sethladd** [comment]
@Hixie is there a good way to link this over to your routing refactoring work, to ensure we unlock this issue as a result of your routing work?

**Comment by Hixie** [comment]
It's not clear to me what exactly is missing to make this possible. Can someone who's tried to do this elaborate on what you have found you need?

**Comment by collinjackson** [comment]
see also: #10884

**Comment by xster** [comment]
@bparrishMines any updates on this?

**Comment by bparrishMines** [comment]
It's essentially done. I'm just having an issue getting a certain method called from the engine. I potentially may have to settle for a workaround for the time being.

**Comment by bparrishMines** [comment]
I finished adding functionality to receive dynamic links. 

https://pub.dartlang.org/packages/firebase_dynamic_links

Should we close this?

**Comment by xster** [comment]
Awesome! Ya

**Comment by kroikie** [comment]
@sethladd

This issue has been moved to https://github.com/FirebaseExtended/flutterfire/issues/975. Any further collaboration will be done there.
